Major update to Tanker

JackDanger opened the pull request:
Ahoy Kidpollo!

Thanks for the fantastic work with Tanker! I really love it, it saved me weeks of work.

I've been using Sunspot and Solr for a year and I just switched to IndexTank. I tried ThinkingTank but it was nowhere near production-ready. Instead I forked Tanker and started adding all the functionality I've grown accustomed to from other search ORM extensions.

What I've added to Tanker:

block-style field definitions:

```
tankit 'index' do
  indexes :author_names do
    authors.map(&:name)
  end
end
```

document variable definitions

```
tankit 'index' do
  variables do
    {0 => id,
     1 => lat,
     2 => lng}
  end
end
```

Multi-class searching:

```
Tanker.search([Topic,User,Post], 'keyword') # returns mixed results of different types
```

Single-class reindexing:

```
Topic.tanker_reindex
```

Oh, and batch indexing. When you run 'rake tanker:reindex' it'll now default to indexing and PUTting 200 records at a time. WAY, WAY, faster than doing it one at a time. I'm indexing 800 records in about 15 seconds.

Automatic timestamping: if the object responds to 'created_at' then that'll be set as the document's creation time in the index.

I also cleaned up the 'tankit' configuration call so it doesn't pollute class-level and object-level methods. I put it into a ModelConfig class so users don't have their Topic.index, Topic.indexes, etc. methods overwritten.

I've added specs for everything and did my best to preserve all of your original organization. I didn't mess with anything that I didn't have to. I've also updated the README.

Care to push a new gem version? :)

edgarjs committed:
Parse found ids to int so ActiveRecord does not complain

JackDanger committed:
Added support for providing values via blocks

Objects don't always have methods corresponding to data they need to
index. This is a feature I've just ported from Sunspot to allow
indexable values from any calculation.

Tests included

JackDanger committed:
using full paths, Ruby 1.9 hates double-loading

JackDanger committed:
writing spec for IndexClient#add_document

Ensuring that the right data is passed in to add_document, not leaving
out anything important.

JackDanger committed:
adding API for dynamic field content to README

JackDanger committed:
adding support for specifying variable data

Spec and README update included in this patch.

I wish IndexTank would let us use string names for these, it's a bit
clumsy to specify these keys as integers.

JackDanger committed:
allowing use with Rails 2.x

This is not an ideal way to backport but, without this patch, Tanker
cannot be included in Rails 2.x projects.

JackDanger committed:
adding automatic Timestamping

This works on all ActiveRecord-like object, easily extensible.

JackDanger committed:
adding batch updating

Added spec and utility that replaces the current rake task logic

JackDanger committed:
adding basic timer to output

JackDanger committed:
fixing typo

JackDanger committed:
getting best performance at 200-count batches

JackDanger committed:
automatically joining arrays

JackDanger committed:
fixing calculation error

JackDanger committed:
not using such a global name for method

JackDanger committed:
not passing in an ignored 'conditions' param

Formerly, this line of code would copy a conditions param into the query
param but let the original options[:conditions] object be serialized and
passed into the indextank.com web service call.
This patch cleans that up.

JackDanger committed:
allowing multiple condition specification

The IndexTank API allows an array of values for the same condition. But
rather than the field:[val1,val2] format it requires a
field:val1+field:val2 format.

This patch automatically translates conditions with array values into
IndexTank API-compatible queries.

JackDanger committed:
Ensuring integer values are integers

Some applications will pass the params[:page] directly into the search
call. Running String#- will raise an error.

JackDanger committed:
fixing bug with setting page '0'

JackDanger committed:
cleaning up redundant code

This same line appears in tanker_index_options

JackDanger committed:
indexing consistently by symbol

This is to ensure we have consistent key format throughout our specs.

JackDanger committed:
using updated syntax for field name expectations

JackDanger committed:
using proper data format for expectations.

JackDanger committed:
spec helper classes have attributes

JackDanger committed:
adding support for querying across models

This feature is ported from Solr's Sunspot. In addition to calling
Person.search_tank("keyword") you can now call Tanker.search([Person,
Dog, Cat], "keyword) and get ordered results from all models.

Eager-loading and result ordering is preserved.
Specs included.

JackDanger committed:
fixing typo

JackDanger committed:
simplifying how conditions are mapped

JackDanger committed:
testing that multi-value phrases work.

If you have an array of values for the same field to specify in your
query IndexTank prefers them itemized.

JackDanger committed:
creating intermediate ModelConfig class

Rather than running index(), indexes(), etc. directly on the model class
(which may need those methods for other purposes), this patch runs those
methods within the scope of a ModelConfig class.

All functionality is identical, the only difference is the model that
includes Tanker has less method namespace pollution

JackDanger committed:
using updated name for variable definition

JackDanger committed:
documenting that you don't need the clunky block param

Field blocks are now executed in the scope of each record.

JackDanger committed:
raise error if tankit() hasn't been added to model

JackDanger committed:
demonstrate that self is record in this context

JackDanger committed:
reliably sorting field data

This is purely to ensure that our specs always have the right field
order. Otherwise the random ordering of hash values causes specs to fail
when they should be passing.

JackDanger committed:
splitting big reindex method into per-class method

Sometimes it's useful to reindex just one class, rather than the entire
data set.
Users can now call Model.tanker_reindex and provide an optional batch
size (defaults to 200)

JackDanger committed:
updating README for new API options

JackDanger committed:
updating README with query variables & batch size

JackDanger committed:
updating call to index_name

This is now stored in the tanker_config object, rather than directly on
the class.

JackDanger committed:
updating rakefile for 'rake spec'

Also runs specs as default task

JackDanger committed:
adding scoping of reindexing

JackDanger committed:
handling queries passed in as lists of words

Just joining them into a single string. No reason to raise an error here
if we can just as easily handle Topic.search(['tag1', 'tag2'])

JackDanger committed:
adding support for index functions

Automatically updated via "rake tanker:functions"

JackDanger committed:
adding myself as an author

kidpollo commented:
Hey JAck!

I greatly appreciate your input, you have no idea how happy I get when I get a pull request! I have been kinda busy and I feel bad I have not integrated your stuff and am happy to continue colaborating with you on this awesome gem :P

I am thingking on promoting this to heroku directly I moved to san francisco and I have some friends there :P

I have not seen the code but did you write specs around your features? If not then it would be cool to have them.

Thanks

PAco

On Mar 29, 2011, at 11:33 PM, JackDanger wrote:

> Ahoy Kidpollo!
> 
> Thanks for the fantastic work with Tanker! I really love it, it saved me weeks of work.
> 
> I've been using Sunspot and Solr for a year and I just switched to IndexTank. I tried ThinkingTank but it was nowhere near production-ready. Instead I forked Tanker and started adding all the functionality I've grown accustomed to from other search ORM extensions.
> 
> What I've added to Tanker:
> 
> block-style field definitions:
> 
>   tankit 'index' do
>     indexes :author_names do
>       authors.map(&:name)
>     end
>   end
> 
> document variable definitions
> 
>   tankit 'index' do
>     variables do
>       {0 => id,
>        1 => lat,
>        2 => lng}
>     end
>   end
> 
> Multi-class searching:
> 
>   Tanker.search([Topic,User,Post], 'keyword') # returns mixed results of different types
> 
> Single-class reindexing:
> 
>   Topic.tanker_reindex
> 
> Oh, and batch indexing. When you run 'rake tanker:reindex' it'll now default to indexing and PUTting 200 records at a time. WAY, WAY, faster than doing it one at a time. I'm indexing 800 records in about 15 seconds.
> 
> Automatic timestamping: if the object responds to 'created_at' then that'll be set as the document's creation time in the index.
> 
> I also cleaned up the 'tankit' configuration call so it doesn't pollute class-level and object-level methods. I put it into a ModelConfig class so users don't have their Topic.index, Topic.indexes, etc. methods overwritten.
> 
> I've added specs for everything and did my best to preserve all of your original organization. I didn't mess with anything that I didn't have to. I've also updated the README.
> 
> Care to push a new gem version? :)
> 
> ## 
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/kidpollo/tanker/pull/3

JackDanger commented:
> I am thingking on promoting this to heroku directly I moved to san
> francisco and I have some friends there :P

Awesome :)

> I have not seen the code but did you write specs around your features? If
> not then it would be cool to have them.

Yup, the only thing that isn't specced is a Rake task for updating
server-side function definitions. But it wouldn't be too hard to add one.
Otherwise it's all tested and working great.

Thanks for doing the hard 90% of the work, it was fun to just add features
on top of a good code base. I started using index_tank and then
thinking_tank and couldn't bear the awful code quality. Tanker was a joyful
find :)

::Jack

> Thanks
> 
> PAco
> 
> On Mar 29, 2011, at 11:33 PM, JackDanger wrote:
> 
> > Ahoy Kidpollo!
> > 
> > Thanks for the fantastic work with Tanker! I really love it, it saved me
> > weeks of work.
> > 
> > I've been using Sunspot and Solr for a year and I just switched to
> > IndexTank. I tried ThinkingTank but it was nowhere near production-ready.
> > Instead I forked Tanker and started adding all the functionality I've grown
> > accustomed to from other search ORM extensions.
> > 
> > What I've added to Tanker:
> > 
> > block-style field definitions:
> > 
> >   tankit 'index' do
> >     indexes :author_names do
> >       authors.map(&:name)
> >     end
> >   end
> > 
> > document variable definitions
> > 
> >   tankit 'index' do
> >     variables do
> >       {0 => id,
> >        1 => lat,
> >        2 => lng}
> >     end
> >   end
> > 
> > Multi-class searching:
> > 
> >   Tanker.search([Topic,User,Post], 'keyword') # returns mixed results of
> > different types
> > 
> > Single-class reindexing:
> > 
> >   Topic.tanker_reindex
> > 
> > Oh, and batch indexing. When you run 'rake tanker:reindex' it'll now
> > default to indexing and PUTting 200 records at a time. WAY, WAY, faster than
> > doing it one at a time. I'm indexing 800 records in about 15 seconds.
> > 
> > Automatic timestamping: if the object responds to 'created_at' then
> > that'll be set as the document's creation time in the index.
> > 
> > I also cleaned up the 'tankit' configuration call so it doesn't pollute
> > class-level and object-level methods. I put it into a ModelConfig class so
> > users don't have their Topic.index, Topic.indexes, etc. methods overwritten.
> > 
> > I've added specs for everything and did my best to preserve all of your
> > original organization. I didn't mess with anything that I didn't have to.
> > I've also updated the README.
> > 
> > Care to push a new gem version? :)
> > 
> > ## 
> > 
> > Reply to this email directly or view it on GitHub:
> > https://github.com/kidpollo/tanker/pull/3
> 
> ## 
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/kidpollo/tanker/pull/3#issuecomment-946220

JackDanger committed:
adding filter_function support

JackDanger committed:
adding support for filter docvars

JackDanger committed:
removing note about alpha, this gem has come a long way

JackDanger committed:
adding README examples for geospatial search

JackDanger committed:
adding note about rake tanker:functions

JackDanger committed:
folks seem to be using this term more

JackDanger commented:
I've added support for filter_functions and filter_docvars (so you can now filter your results to just X miles from some point) and I've updated the README with an example of a fulltext geospatial search. The new code is working, tested, and live on my company site at http://cloops.com/search

kidpollo commented on README.rdoc:
```diff
@@ -50,7 +49,32 @@ in the IndexTank Dashboard
         indexes :tag_list #NOTICE this is an array of Tags! Awesome!
+
+        # you may also dynamically retrieve field data
+        indexes :author_names do
```
this is awesome!
